src/ufuncs.c

```diff
@@ -1108,6 +1108,79 @@ static void buffer_func(char** args, npy_intp* dimensions, npy_intp* steps, void
 }
 static PyUFuncGenericFunction buffer_funcs[1] = {&buffer_func};
 
+static char offset_curve_dtypes[6] = {NPY_OBJECT, NPY_DOUBLE, NPY_INT,
+                                      NPY_INT,    NPY_DOUBLE, NPY_OBJECT};
+static void offset_curve_func(char** args, npy_intp* dimensions, npy_intp* steps,
+                              void* data) {
+  char *ip1 = args[0], *ip2 = args[1], *ip3 = args[2], *ip4 = args[3], *ip5 = args[4];
+  npy_intp is1 = steps[0], is2 = steps[1], is3 = steps[2], is4 = steps[3], is5 = steps[4];
+  npy_intp n = dimensions[0];
+  npy_intp i;
+  GEOSGeometry** geom_arr;
+  GEOSGeometry* in1 = NULL;
+
+  // Fail if inputs output multiple times on the same place in memory. That would
+  // lead to segfaults as the same GEOSGeometry would be 'owned' by multiple PyObjects.
+  if ((steps[5] == 0) && (dimensions[0] > 1)) {
+    PyErr_Format(PyExc_NotImplementedError,
+                 "Unknown ufunc mode with args[0]=%p, args[5]=%p, steps[0]=%ld, "
+                 "steps[5]=%ld, dimensions[0]=%ld.",
+                 args[0], args[5], steps[0], steps[5], dimensions[0]);
+    return;
+  }
+
+  if ((is3 != 0) | (is4 != 0) | (is5 != 0)) {
+    PyErr_Format(PyExc_ValueError,
+                 "Offset curve function called with non-scalar parameters");
+    return;
+  }
+
+  double width;
+  int quadsegs = *(int*)ip3;
+  int joinStyle = *(int*)ip4;
+  double mitreLimit = *(double*)ip5;
+
+  // allocate a temporary array to store output GEOSGeometry objects
+  geom_arr = malloc(sizeof(void*) * n);
+  if (geom_arr == NULL) {
+    PyErr_SetString(PyExc_MemoryError, "Could not allocate memory");
+    return;
+  }
+
+  GEOS_INIT_THREADS;
+
+  for (i = 0; i < n; i++, ip1 += is1, ip2 += is2) {
+    /* get the geometry: return on error */
+    if (!get_geom(*(GeometryObject**)ip1, &in1)) {
+      errstate = PGERR_NOT_A_GEOMETRY;
+      destroy_geom_arr(ctx, geom_arr, i - 1);
+      break;
+    }
+
+    width = *(double*)ip2;
+    if ((in1 == NULL) | npy_isnan(width)) {
```

Comment:
One note here: GEOS actually can "handle" the width being NaN, in the sense that it will raise a proper GEOSException. 
It could also make sense to raise an exception here, as I am not sure there are good use case of the width being NaN (so this doesn't go silent)